File: auntieos-admin/src/lib/pendingAddKinfolk.ts
import type { NewKinfolkInput } from '../api/directoryWrite';
import type { EmergencyContactDraft } from '../api/emergencyContacts';

/**
 * #890: a household Add Kinfolk created whose Emergency Contact has not saved yet.
 *
 * Add creates the household, then saves its contact. When the contact save
 * failed and the operator closed the dialog, the created id used to live only in
 * the dialog's state and was gone, so the next Add made a second household. This
 * keeps it outside the dialog, per operator, for the rest of the browser session,
 * until the contact saves or the operator chooses Discard. Admin Android keeps the
 * same thing in DirectoryViewModel (`leaveAddKinfolk`).
 *
 * SESSION STORAGE, with a copy in memory. Storage carries it across a reload
 * within the tab. Every access is wrapped because storage throws outright in some
 * privacy modes, and then the in-memory copy still carries it for this page.
 *
 * DISCARD writes nothing to the household. It records the discarded id (#907
 * review item 1a), because Discard means "the next Add is a new household": the
 * next create sends that id as `ignoreDuplicateOf`, so the server's duplicate
 * check does not hand the discarded household back.
 *
 * A DUPLICATE ANSWER (#907 review item 1b). When the server answers `duplicateOf`,
 * what the operator typed is kept here, keyed by that household, so its edit
 * screen can fill in the fields that differ as unsaved changes. Nothing typed is
 * lost and nothing is written until the operator saves.
 */

export const PENDING_ADD_KINFOLK_STORAGE_PREFIX = 'auntieos.pendingAddKinfolk.';
export const DISCARDED_ADD_KINFOLK_STORAGE_PREFIX = 'auntieos.discardedAddKinfolk.';
export const DUPLICATE_ADD_KINFOLK_STORAGE_PREFIX = 'auntieos.duplicateAddKinfolk.';

export interface PendingAddKinfolk {
  kinfolkId: string;
  /** The household fields as they were saved, shown locked when Add continues. */
  household: NewKinfolkInput;
  /** The contact as last typed, so continuing does not make the operator retype it. */
  contacts: EmergencyContactDraft[];
}

/** What the operator typed into an Add the server answered with `duplicateOf`. Same shape. */
export type DuplicateAddKinfolk = PendingAddKinfolk;

const memory = new Map<string, unknown>();

function isPending(v: unknown): v is PendingAddKinfolk {
  if (!v || typeof v !== 'object') return false;
  const o = v as Record<string, unknown>;
  return (
    typeof o['kinfolkId'] === 'string' &&
    o['kinfolkId'] !== '' &&
    !!o['household'] &&
    typeof o['household'] === 'object' &&
    Array.isArray(o['contacts'])
  );
}

function isId(v: unknown): v is string {
  return typeof v === 'string' && v !== '';
}

function load<T>(key: string, valid: (v: unknown) => v is T): T | null {
  const fromMemory = (): T | null => {
    const m = memory.get(key);
    return valid(m) ? m : null;
  };
  try {
    const raw = sessionStorage.getItem(key);
    if (raw !== null) {
      const parsed: unknown = JSON.parse(raw);
      return valid(parsed) ? parsed : null;
    }
    // Nothing stored: either there is nothing kept, or storage refused the write.
    return fromMemory();
  } catch {
    return fromMemory();
  }
}

function store(key: string, value: unknown): void {
  memory.set(key, value);
  try {
    sessionStorage.setItem(key, JSON.stringify(value));
  } catch {
    // Storage refused. The in-memory copy still carries it for this page.
  }
}

function remove(key: string): void {
  memory.delete(key);
  try {
    sessionStorage.removeItem(key);
  } catch {
    // Nothing stored to remove.
  }
}

export function readPendingAddKinfolk(uid: string | null | undefined): PendingAddKinfolk | null {
  if (!uid) return null;
  return load(`${PENDING_ADD_KINFOLK_STORAGE_PREFIX}${uid}`, isPending);
}

export function savePendingAddKinfolk(uid: string | null | undefined, pending: PendingAddKinfolk): void {
  if (!uid) return;
  store(`${PENDING_ADD_KINFOLK_STORAGE_PREFIX}${uid}`, pending);
}

export function clearPendingAddKinfolk(uid: string | null | undefined): void {
  if (!uid) return;
  remove(`${PENDING_ADD_KINFOLK_STORAGE_PREFIX}${uid}`);
}

/** The household this operator last Discarded, sent as `ignoreDuplicateOf` on their next create. */
export function readDiscardedAddKinfolk(uid: string | null | undefined): string | null {
  if (!uid) return null;
  return load(`${DISCARDED_ADD_KINFOLK_STORAGE_PREFIX}${uid}`, isId);
}

export function saveDiscardedAddKinfolk(uid: string | null | undefined, kinfolkId: string): void {
  if (!uid || kinfolkId === '') return;
  store(`${DISCARDED_ADD_KINFOLK_STORAGE_PREFIX}${uid}`, kinfolkId);
}

export function clearDiscardedAddKinfolk(uid: string | null | undefined): void {
  if (!uid) return;
  remove(`${DISCARDED_ADD_KINFOLK_STORAGE_PREFIX}${uid}`);
}

/** Kept for the household's edit screen. One at a time per operator: a newer duplicate replaces it. */
export function saveDuplicateAddKinfolk(uid: string | null | undefined, typed: DuplicateAddKinfolk): void {
  if (!uid) return;
  store(`${DUPLICATE_ADD_KINFOLK_STORAGE_PREFIX}${uid}`, typed);
}

/** What was typed for THIS household, or null. Another household's edit screen never sees it. */
export function readDuplicateAddKinfolk(uid: string | null | undefined, kinfolkId: string): DuplicateAddKinfolk | null {
  if (!uid) return null;
  const typed = load(`${DUPLICATE_ADD_KINFOLK_STORAGE_PREFIX}${uid}`, isPending);
  return typed !== null && typed.kinfolkId === kinfolkId ? typed : null;
}

export function clearDuplicateAddKinfolk(uid: string | null | undefined): void {
  if (!uid) return;
  remove(`${DUPLICATE_ADD_KINFOLK_STORAGE_PREFIX}${uid}`);
}

/** How the Continue prompt names the household. */
export function pendingHouseholdName(pending: PendingAddKinfolk): string {
  const name = `${pending.household.firstName.trim()} ${pending.household.lastName.trim()}`.trim();
  return name === '' ? 'this household' : name;
}

File: auntieos-admin/web/composeApp/src/commonMain/kotlin/com/tribetails/auntieos/web/screens/directory/PendingAddKinfolk.kt
package com.tribetails.auntieos.web.screens.directory

import androidx.compose.runtime.mutableStateMapOf
import com.tribetails.auntieos.web.data.EmergencyContactDraft
import com.tribetails.auntieos.web.data.Kinfolk

/**
 * #890: a household Add Kinfolk created whose Emergency Contact has not saved yet.
 *
 * Add creates the household, then saves its contact. When the contact save failed
 * and the operator left the screen, the created id used to live only in
 * [KinfolkEditScreen]'s state and was gone, so the next Add made a second
 * household. This keeps it at app level, per operator, for as long as the console
 * runs, until the contact saves or the operator chooses Discard. Admin Android
 * keeps the same thing in DirectoryViewModel (`leaveAddKinfolk`); admin web keeps
 * it in session storage (`lib/pendingAddKinfolk.ts`).
 *
 * [household] is the form as it was saved, so continuing shows the same locked
 * fields; [contacts] is the contact as last typed.
 */
data class PendingKinfolk(
    val kinfolkId: String,
    val household: Kinfolk,
    val contacts: List<EmergencyContactDraft>,
)

object PendingAddKinfolk {
    /** Compose state, so a screen reading it recomposes when Discard clears it. Keyed by operator uid. */
    private val byOperator = mutableStateMapOf<String, PendingKinfolk>()

    fun get(operatorUid: String?): PendingKinfolk? = byOperator[operatorUid.orEmpty()]

    fun keep(operatorUid: String?, pending: PendingKinfolk) {
        byOperator[operatorUid.orEmpty()] = pending
    }

    fun clear(operatorUid: String?) {
        byOperator.remove(operatorUid.orEmpty())
    }

    /**
     * #907 review item 1(a): the household each operator last Discarded. Discard
     * says the next Add is a new household, so the next create sends this id as
     * `ignoreDuplicateOf` and the server's duplicate check skips it.
     */
    private val discardedByOperator = mutableStateMapOf<String, String>()

    fun discard(operatorUid: String?, kinfolkId: String) {
        if (kinfolkId.isNotBlank()) discardedByOperator[operatorUid.orEmpty()] = kinfolkId
    }

    fun discardedFor(operatorUid: String?): String? = discardedByOperator[operatorUid.orEmpty()]

    fun clearDiscarded(operatorUid: String?) {
        discardedByOperator.remove(operatorUid.orEmpty())
    }

    /**
     * #907 review item 1(b): what an operator typed into an Add the server answered
     * with `duplicateOf`, kept for that household's edit screen. One per operator.
     */
    private val duplicateByOperator = mutableStateMapOf<String, PendingKinfolk>()

    fun keepDuplicate(operatorUid: String?, typed: PendingKinfolk) {
        duplicateByOperator[operatorUid.orEmpty()] = typed
    }

    /** The typing for THIS household only; another household's edit screen never sees it. */
    fun duplicateFor(operatorUid: String?, kinfolkId: String): PendingKinfolk? =
        duplicateByOperator[operatorUid.orEmpty()]?.takeIf { it.kinfolkId == kinfolkId }

    fun clearDuplicate(operatorUid: String?) {
        duplicateByOperator.remove(operatorUid.orEmpty())
    }

    /** Test seam: forget every operator's pending, discarded and duplicate households. */
    fun clearAll() {
        byOperator.clear()
        discardedByOperator.clear()
        duplicateByOperator.clear()
    }
}

/**
 * #907 review item 1(b): the stored household with a duplicate Add's typing laid
 * over it. A typed value counts only when it is not blank and differs from the
 * stored one once both are trimmed; a blank is "not typed", never "clear it". The
 * id, status and everything Add has no field for stay as stored.
 */
fun overlayDuplicateAdd(stored: Kinfolk, typed: Kinfolk): Kinfolk {
    fun pick(t: String, s: String): String = if (t.isNotBlank() && t.trim() != s.trim()) t.trim() else s
    return stored.copy(
        firstName           = pick(typed.firstName, stored.firstName),
        lastName            = pick(typed.lastName, stored.lastName),
        phoneNumber         = pick(typed.phoneNumber, stored.phoneNumber),
        secondaryPhone      = pick(typed.secondaryPhone, stored.secondaryPhone),
        email               = pick(typed.email, stored.email),
        secondaryEmail      = pick(typed.secondaryEmail, stored.secondaryEmail),
        serviceAddress      = pick(typed.serviceAddress, stored.serviceAddress),
        gateCode            = pick(typed.gateCode, stored.gateCode),
        parkingInstructions = pick(typed.parkingInstructions, stored.parkingInstructions),
        entryNotes          = pick(typed.entryNotes, stored.entryNotes),
        wifiName            = pick(typed.wifiName, stored.wifiName),
        wifiPassword        = pick(typed.wifiPassword, stored.wifiPassword),
        internalNotes       = pick(typed.internalNotes, stored.internalNotes),
        referralSource      = pick(typed.referralSource, stored.referralSource),
        vetClinicName       = pick(typed.vetClinicName, stored.vetClinicName),
        vetClinicPhone      = pick(typed.vetClinicPhone, stored.vetClinicPhone),
        vetClinicAddress    = pick(typed.vetClinicAddress, stored.vetClinicAddress),
        formValues          = stored.formValues + typed.formValues.filter { (k, v) ->
            v.isNotBlank() && v.trim() != stored.formValues[k].orEmpty().trim()
        }.mapValues { it.value.trim() },
    )
}

/**
 * #907 review item 1(b): the edit screen's notice. The desktop edit screen has no
 * status control (status is set on Add only), so a status typed on Add that differs
 * is named here rather than filled in.
 */
fun duplicateAddNotice(stored: Kinfolk, typed: Kinfolk): String {
    val name = "${stored.firstName.trim()} ${stored.lastName.trim()}".trim().ifBlank { "This household" }
    val notice = "$name was already added a few minutes ago. What you typed in Add that differs is filled in below and is not saved yet."
    val typedStatus = typed.status.trim()
    return if (typedStatus.isNotBlank() && !typedStatus.equals(stored.status.trim(), ignoreCase = true)) {
        "$notice Status on Add was $typedStatus; this household is ${stored.status.trim().ifBlank { "blank" }}, and status is not changed here."
    } else notice
}

/** How the Continue prompt names the household. */
fun pendingHouseholdName(pending: PendingKinfolk): String =
    "${pending.household.firstName.trim()} ${pending.household.lastName.trim()}".trim().ifBlank { "this household" }
